server/src/main/java/io/crate/expression/tablefunctions/MatchesFunction.java

```diff
@@ -0,0 +1,236 @@
+        }
+        String flags = null;
+        if (arguments.size() == 3) {
+            assert arguments.get(2).symbolType() == SymbolType.LITERAL :
+                "3rd argument must be a " + SymbolType.LITERAL;
+            flags = ((Literal<String>) arguments.get(2)).value();
+        }
+        if (pattern != null) {
+            return new MatchesFunction(
+                signature, boundSignature, createMatcher(pattern, flags));
+        }
+        return this;
+    }
+
+    @Override
+    public Iterable<Row> evaluate(TransactionContext txnCtx, Input[] args) {
+        assert args.length == 2 || args.length == 3 : "number of args must be 2 or 3";
+        String value = (String) args[0].value();
+        String pattern = (String) args[1].value();
+        if (value == null || pattern == null) {
+            return () -> NO_RESULTS_ITERATOR;
+        }
+
+        RegexMatcher matcher;
+        if (regexMatcher == null) {
+            String flags = null;
+            if (args.length == 3) {
+                flags = (String) args[2].value();
+            }
+            matcher = createMatcher(pattern, flags);
+        } else {
+            matcher = regexMatcher;
+        }
+
+        Iterator<Row> rowIterator;
+        List<List<String>> rowGroups = matcher.match(value);
+        if (rowGroups == null) {
+            rowIterator = NO_RESULTS_ITERATOR;
+        } else {
+            rowIterator = new Iterator<>() {
+
+                final Object [] columns = new Object[]{ null };
+                final RowN row = new RowN(columns);
+                int idx = 0;
+
+                @Override
+                public boolean hasNext() {
+                    return idx < rowGroups.size();
+                }
+
+                @Override
+                public Row next() {
+                    if (!hasNext()) {
+                        throw new NoSuchElementException("no more rows");
+                    }
+                    columns[0] = rowGroups.get(idx++);
+                    return row;
+                }
+            };
+        }
+        return () -> rowIterator;
+    }
+
+    private static RegexMatcher createMatcher(String regex, @Nullable String flags) {
+
+        Function<Supplier<List<String>>, List<List<String>>> collector =
+            RegexMatcher.isGlobal(flags) ?
+                (supplier) -> {
+                    List<List<String>> result = Stream
+                        .generate(supplier)
+                        .takeWhile(Objects::nonNull)
+                        .collect(Collectors.toList());
+                    return result.isEmpty() ? null : result;
+                }
+                :
+                (supplier) -> {
+                    List<String> groups = supplier.get();
+                    return groups != null ? List.of(groups) : null;
+                };
+        return new RegexMatcher(regex, flags) {
+
+            @Override
+            public List<List<String>> match(String term) {
```

Comment:
maybe we can make this function nonnullable, then we probably won't have to handle special null cases here https://github.com/crate/crate/pull/10104/files#diff-4f4a1a2d0baa1a4764044bf3363dcba8R168 and use the same logic to handle both empty and non-empty cases?